core/shared/src/main/scala/zio/ZLayer.scala

```diff
@@ -98,18 +98,47 @@ final class ZLayer[-RIn, +E, +ROut <: Has[_]] private (
   def fresh: ZLayer[RIn, E, ROut] =
     new ZLayer(self.scope)
 
+  /**
+   * Returns a new layer whose output is mapped by the specified function.
+   */
+  def map[ROut1 >: ROut <: Has[_]](f: ROut => ROut1): ZLayer[RIn, E, ROut1] =
+    self >>> ZLayer.fromEnvironment(f)
+
+  /**
+   * Returns a layer with its error channel mapped using the specified
+   * function.
+   */
+  def mapError[E1 >: E](f: E => E1): ZLayer[RIn, E1, ROut] =
+    new ZLayer(scope.map(run => run(_).mapError(f)))
+
+  /**
+   * Overwrites the services output by this laywr with the services output by
+   * the specified layer.
+   */
+  def overwrite[E1 >: E, RIn2, A: Tagged](
+    that: ZLayer[RIn2, E1, Has[A]]
+  )(implicit ev: ROut <:< Has[A]): ZLayer[RIn with RIn2, E1, ROut] =
+    new ZLayer(
+      Managed.finalizerRef(_ => UIO.unit).map { finalizerRef => memoMap =>
+        for {
+          l <- memoMap.getOrElseMemoize(self, finalizerRef)
+          r <- memoMap.getOrElseMemoize(that, finalizerRef)
+        } yield l.update[A](_ => r.get[A])
```

Comment:
I noticed (like you pointed out) this is a specialization fo `++` when `ROut2` is just a single `Has` that is already a part of `ROut`.

The implementation is almost the same, except instead of `union`, `update` is being called to do a targeted replacement of `A`.

I wonder why `union` would not overwrite in this case? In theory, `l union r` should always cause `r` to overwrite any parts already contained by `l`.

Any thoughts on this?

Reply:
I agree. I haven't done a deep dive on this yet. Maybe something to do with the ordering of the underlying map implementations? I can look into this.

Reply:
It looks like when `l` and `r` are combined with `union` in `++` we get a `Has` instance that has two entries in the map associated with the `Clock`, which is leading to the indeterminacy.

Reply:
More updates on this. I think the issue is that in addition to having a string representation type tags have a hash code and the hash code is not always the same for the same type. In particular, it appears that if a type tag is summoned from different compilation units the hash code will be different. So for example if I print `implicitly[Tagged[Has[Clock.Service]]]` from `zio.clock` and `zio.test.environment.EnvironmentSpec` they will have different hash codes. That's why we're getting duplicate entries in the map underlying `Has`. When we do `unionAll` in `Has` we're assuming that we will just overwrite the old implementations of the same service but instead we're adding a second one.

I think we probably want to say that two tags are equal if they have are for the same type so I think we probably either need our own type to wrap tags that hashes them based on their string representation or a specialized map for type tags that hashes them based on their string representation.

Reply:
Oh my. Well, Kai did warn us. 😆  

I think we absolutely will replace `TypeTag` but until then we could wrap it in something like:

```scala
case class FixedTypeTag[A](value: TypeTag[A]) {
  override def equals(that: Any): Boolean = 
    that match { case FixedTypeTag(that) => value.toString == that.toString }
  override def hashCode = value.toString.hashCode
}
```

Then `Tagged[A]` becomes `FixedTypeTag[A]` and no ZIO code (outside that package object) needs to know how broken it is. Obviously we gotta fix the problem at the source for RC18.